File: bitcoin-stockmem-ts/src/data/label-generator.ts
/**
 * Generate up/down/flat labels from daily price returns.
 */

import { PRICE_THRESHOLD } from "../config";
import type { PriceRow } from "./price-fetcher";

export interface LabelledRow extends PriceRow {
  next_return: number | null;
  label: "up" | "down" | "flat";
}

export function generateLabels(
  rows: PriceRow[],
  threshold = PRICE_THRESHOLD
): LabelledRow[] {
  return rows.map((row, i) => {
    const nextReturn = i < rows.length - 1 ? rows[i + 1].return_pct : null;
    let label: "up" | "down" | "flat" = "flat";
    if (nextReturn !== null && nextReturn !== undefined) {
      if (nextReturn > threshold) label = "up";
      else if (nextReturn < -threshold) label = "down";
    }
    return { ...row, next_return: nextReturn, label };
  });
}

export function filterTradableDays(rows: LabelledRow[]): LabelledRow[] {
  return rows.filter((r) => r.label !== "flat");
}

File: bitcoin_stockmem/data/label_generator.py
"""Generate up/down/flat labels from daily price returns."""

from __future__ import annotations

import pandas as pd

from config import PRICE_THRESHOLD


def generate_labels(
    prices_df: pd.DataFrame,
    threshold: float | None = None,
) -> pd.DataFrame:
    """Add a 'label' column to a price DataFrame.

    Args:
        prices_df: Must have columns 'date' and 'return_pct'.
        threshold: Override the default ±2% threshold.

    Returns:
        Same DataFrame with added 'label' column ("up", "down", or "flat").
        The label for day t reflects the return from t to t+1.
    """
    thr = threshold or PRICE_THRESHOLD
    df = prices_df.copy()

    # Shift return by -1: label[t] = return from t to t+1
    df["next_return"] = df["return_pct"].shift(-1)

    conditions = [
        df["next_return"] > thr,
        df["next_return"] < -thr,
    ]
    choices = ["up", "down"]
    df["label"] = pd.np.select(conditions, choices, default="flat") if hasattr(pd, "np") else "flat"

    # Use numpy directly for compatibility
    import numpy as np
    df["label"] = np.select(conditions, choices, default="flat")

    return df


def filter_tradable_days(df: pd.DataFrame) -> pd.DataFrame:
    """Remove 'flat' days from evaluation (following the paper)."""
    return df[df["label"] != "flat"].copy()
